README.md
# random-words

Generate new, unique words from a seed set of words

## USAGE

    random-words [seed.txt]
    
print a markov chain generated list of random words

## OPTIONS

### -c, --count [number]      

maximum number of unique words to generate (default 100)

### -f, --frequency
    
show frequency counts for each unique word

### -i, --influence
    
each input line may have a number, the influence of words on that line

### -l, --limit [number]
    
limit output to how many most frequent words (default all)

### -m, --minimum [number]

minimum length of word to generate (default, and lowest, is window size)

### -w, --window [number]

size of the markov analysis window in characters, (2+, default 3)


## EXAMPLE

    random-words -c 50 elements.txt

    Francient Plutetium Borgium Berkel Siliconium Platinium Pallium 
    Arsenium Chlorium Ancium Actinum Radolinium Sodine Amercury Meitnericium 
    Iodium Prosium Cericium Strogen Kryptonium Brominum Einsten Thorine 
    Thalladium Merium Selevium Bromethium Thorus Tanium Alumine Lantaluminum 
    Dysprotactinium Amerium Seaboron Bromium Magnese Anthallium Fermanganese 
    Antium Lawrencient Manganesium Nitron Zircon Lantium Dyspromethium 
    Actine Alum Rhodine Rutherfornium Titantimony


## NOTES

Results are printed to STDOUT, separated by spaces. 

STDIN is not accepted, you must supply a file. 

Duplicates, and words that appear in the orginal text are not included in the output. 

If the probability of generating unique words is low, it may take a while to generate a large set of unique words.

Project is written in SAI and transpiled to Javascript. For more about this, see npm/sai-language



changelog.md
# random-words changelog

## 1.0.5

 - syntax updates for sai 0.2.0
 

## 1.0.4

 - Minor syntax updates to bring up to date with sai 0.1.22
 

## 1.0.3

 - added changelog
 - minor changes for SAI 0.1.19
 

MarkovWord.js
#!/usr/bin/env node

// Javascript source for MarkovWord transpiled by SAI
//

"use strict";

var prototype=new function() {
this.Constructor=function(){};
this.__tobelocked=[];
this.__tobefrozen=[];
this.__contracts=[];
this.__unverified=true;
this.isof={};
return this;
}();
var $AI=require("sai-library");
// Generated code follows
var __context={"name":"MarkovWord","loader":"SAI.GetSourceFromPaths","path":"MarkovWord.sai","mtime":"2018-03-15T01:09:16.413Z","fetched":"2018-03-15T23:05:40.829Z"};
var _Floor = Math.floor;
var _Random = Math.random;
var _WATCHDOG = 100;
var isa = prototype.isa = 'MarkovWord';
var $bindfail = function(m) {
  throw new Error("SAI: A call to " + m + " on object " + isa + " has come unbound from any instance of that object. (If this is not in error, mark the declaration of " + m + " as unbound.)");
}
prototype.isof['MarkovWord'] = {
  version: '0.0.0-unspecified',
  isa: isa,
  context: __context,
  type: "main"
};
prototype.__tobelocked = prototype.__tobelocked.concat(["success", "window", "Configure", "Finalize", "AnalyzeUnweighted", "AnalyzeWeighted", "Generate", "AddWord", "GetWord", "isa"]);
prototype.__tobefrozen = prototype.__tobefrozen.concat(["isof"]);
var $1g = function() {
  var $ = this;
  return {
    window: 3,
    frequency: false,
    minimum: 0,
    settings: ['window', 'frequency', 'minimum'],
    leader: '',
    corpus: [],
    markov: {},
    totals: {},
    _window: undefined
  };
}
for (var i in $1g()) {
  prototype[i] = undefined;
};
var $2g = prototype.Constructor || function() {};
prototype.Constructor = function() {
  $2g.call(this);
  var t = $1g();
  for (var i in t)
    if (t[i] !== undefined) this[i] = t[i];
};
Object.defineProperty(prototype, "success", {
  configurable: true,
  enumerable: true,
  get: function(p) {
    var $ = this; {
      return (($.corpus.length > 0));
    }
  },
  set: undefined
});
Object.defineProperty(prototype, "window", {
  configurable: true,
  enumerable: true,
  get: function(p) {
    var $0, $ = this; {
      return ($._window);
    }
  },
  set: function(p) {
    var $_d, $_c, $1, _w = p,
      $ = this; {
      $._window = _w;
      $.leader = '_'.repeat(_w);
      $.corpus = [];
      $.markov = {};
      $.totals = {};
    }
  }
});
var $6g = prototype['Configure'] || function() {};
prototype['Configure'] = function(p) {
  var $7, $6, $5, $4, $3, $2, _options = p,
    $ = this['Configure'] ? this : $bindfail('Configure'); {
    var $5 = $AI.kviterate_op($AI.select_op(_options, $.settings));
    var $6 = $5.next();
    if (!$6.done)
      do {
        $3 = $6.value[1];
        $4 = $6.value[0];
        $[$4] = $3;
        $6 = $5.next();
      } while (!$6.done);
  }
};
var $7g = prototype['Finalize'] || function() {};
prototype['Finalize'] = function(p) {
  var $12, $11, $10, $9, $8, $ = this['Finalize'] ? this : $bindfail('Finalize'); {
    var $11 = $.markov,
      $10 = undefined;
    if (undefined === $11) $11 = {};
    for ($10 in $11) {
      $9 = $11[$10];
      $.totals[$10] = $AI.reduce_op_fast($9, function(s, i) {
        return (undefined === s) ? i : s + i;
      }, undefined);;
    }
  }
};
var $8g = prototype['AnalyzeUnweighted'] || function() {};
prototype['AnalyzeUnweighted'] = function(p) {
  var $_d, $_c, $22, $21, $20, $19, $18, $17, $16, $15, $14, $13, _source = p,
    $ = this['AnalyzeUnweighted'] ? this : $bindfail('AnalyzeUnweighted'); {
    $.corpus = $AI.filter_op(((undefined === ($_c = (($_d = ((undefined === ($_c = (($_d = ((undefined === ($_c = (($_d = _source).toLowerCase()))) ? $_d : $_c)).replace(/[^a-z\-]/g, ' ')))) ? $_d : $_c)).split(/\s+/)))) ? $_d : $_c), function($14, $15) {
      return ($14.length >= $.window);
    });
    var $20 = $AI.kviterate_op($AI.filter_op($.corpus, function($16, $17) {
      return ($16.length > $.window);
    }));
    var $21 = $20.next();
    if (!$21.done)
      do {
        $18 = $21.value[1];
        $19 = $21.value[0];
        $.AddWord($18, 1);
        $21 = $20.next();
      } while (!$21.done);
    $.Finalize();
  }
};
var $9g = prototype['AnalyzeWeighted'] || function() {};
prototype['AnalyzeWeighted'] = function(p) {
  var $_d, $_c, $_b, $38, $37, $36, $35, $34, $33, $32, $31, $30, $29, $28, _word, _weight, _line, $27, $26, $25, $24, _valid, $23, _source = p,
    $ = this['AnalyzeWeighted'] ? this : $bindfail('AnalyzeWeighted'); {
    _valid = [];
    $.corpus = $AI.sort_op($AI.map_op($AI.map_op(((undefined === ($_c = (($_d = ((undefined === ($_c = (($_d = ((undefined === ($_c = (($_d = _source).toLowerCase()))) ? $_d : $_c)).replace(/[^a-z\-0-9]/g, ' ')))) ? $_d : $_c)).split(/[\r\n]/)))) ? $_d : $_c), function($24, $25) {
      return $24.split(/\s+/);
    }), function($26, $27) {
      return $26;
    }), function(a, b) {
      var aa, bb;
      aa = ($AI.number_op(a));
      bb = ($AI.number_op(b));
      if (aa < bb) return 1;
      if (aa > bb) return -1;;
      return 0;
    });
    var $36 = $.corpus,
      $35;
    if (undefined === $36) $36 = [];
    var $37 = $36.length;
    for ($35 = 0; $35 < $37; $35++) {
      _line = $36[$35];
      _weight = 1;
      var $32 = _line,
        $31;
      if (undefined === $32) $32 = [];
      var $33 = $32.length;
      for ($31 = 0; $31 < $33; $31++) {
        _word = $32[$31];
        if ($29 = (($AI.number_op(_word)))) {
          _weight = $29;
        } else if ($30 = ((_word.length >= $.window))) {
          _valid.push(_word);
          if ($28 = ((_word.length > $.window))) {
            $.AddWord(_word, _weight);
          }
        };
      };
    }
    $.corpus = ((undefined !== ($_b = _valid)) ? $_b : {});
    $.Finalize();
  }
};
var $10g = prototype['Generate'] || function() {};
prototype['Generate'] = function(p) {
  var $_d, $_c, $47, $46, $45, $44, $43, _word, _duplicate, $42, $41, $40, _tally, _results, $39, _wordcount = p,
    $ = this['Generate'] ? this : $bindfail('Generate'); {
    _results = [];
    _tally = $AI.reduce_op_fast($.corpus, function($42, $40, $41) {
      $42[$40] = 1;;
      return $42;
    }, {});
    _duplicate = (_WATCHDOG * _wordcount);
    var $43;
    while ($43 = (((_results.length < _wordcount) && (_duplicate > 0)))) {
      _word = $.GetWord();
      if (!((_tally[_word] || (_word.length < $.minimum)))) {
        _results.push(_word);
      } else {
        _duplicate--;
      }
      _tally[_word] = (_tally[_word] || 0) + 1;
    }
    _results = $AI.sort_op(_results, function(a, b) {
      var aa, bb;
      aa = _tally[a];
      bb = _tally[b];
      if (aa < bb) return 1;
      if (aa > bb) return -1;;
      return 0;
    });
    if ($47 = ($.frequency)) {
      $46 = _results;
      _results = $AI.map_op($46, function($44, $45) {
        return '' + $44 + '(' + _tally[$44] + ')';
      });
    }
    return (_results);
  }
};
var $11g = prototype['AddWord'] || function() {};
prototype['AddWord'] = function(p, _weight) {
  var $_b, $52, $51, $50, $49, _predicted, _current, $48, _word = p,
    $ = this['AddWord'] ? this : $bindfail('AddWord'); {
    _word = '' + $.leader + _word + '_';
    var $49 = 0,
      $50 = (_word.length - $.window),
      $51 = 1;
    if ($51 <= 0) throw new Error("SAI Runtime: COUNT STEP value should be positive.");
    for (; $49 < $50; $49 = $49 + $51) {
      _current = $AI.slice_op(_word, $49, $.window);
      _predicted = $AI.slice_op(_word, ($49 + $.window), 1);
      if (undefined === $.markov[_current]) $.markov[_current] = {};
      $.markov[_current][_predicted] = (_weight + ((undefined !== ($_b = $.markov[_current][_predicted])) ? $_b : 0));;
    }
  }
};
var $12g = prototype['GetWord'] || function() {};
prototype['GetWord'] = function(p) {
  var $_d, $_c, $58, $57, $56, $55, $54, _predicted, _selection, _current, _word, $53, $ = this['GetWord'] ? this : $bindfail('GetWord'); {
    _word = $.leader;
    do {
      _current = $AI.slice_op(_word, undefined, (0 - ($.window)));
      _selection = _Floor(($.totals[_current] * _Random()));
      _predicted = $AI.reduce_op_fast($.markov[_current], function($58, $56, $57) {
        if ($55 = ((_selection >= 0))) {
          _selection = (_selection - $56);
          if ($54 = ((_selection < 0))) {
            $58 = $57;
          }
        };
        return $58;
      }, 'BAD_MARKOV');
      _word += _predicted;
    } while (!((_predicted === '_')));
    _word = $AI.slice_op(_word, $.window, (0 - (1)));
    return (_word);
  }
};
$AI.finalizePrototype(prototype);
if (prototype.isof[prototype.isa].type === "singleton") { prototype.Constructor(); prototype.Instantiate(); }
var result=prototype.isof[prototype.isa].type === "main" ? prototype.constructor() : prototype;
exports=result; try { module.exports=result; } catch(e) {};
return result;


random-words.js
#!/usr/bin/env node

// Javascript source for random-words transpiled by SAI
//

"use strict";

var prototype=new function() {
this.Constructor=function(){};
this.__tobelocked=[];
this.__tobefrozen=[];
this.__contracts=[];
this.__unverified=true;
this.isof={};
return this;
}();
var $AI=require("sai-library");
// Generated code follows
var __context={"name":"random-words","loader":"SAI.GetSourceFromPaths","path":"random-words.sai","mtime":"2018-03-15T01:05:09.667Z","fetched":"2018-03-15T23:05:40.723Z"};
var _FS = require('fs');
var _ReadFile = _FS.readFileSync;
var _ARGV = process.argv;
var _Exit = process.exit;
var _WATCHDOG = 100;
var isa = prototype.isa = 'RW';
var $bindfail = function(m) {
  throw new Error("SAI: A call to " + m + " on object " + isa + " has come unbound from any instance of that object. (If this is not in error, mark the declaration of " + m + " as unbound.)");
}
prototype.isof['RW'] = {
  version: '0.0.0-unspecified',
  isa: isa,
  context: __context,
  type: "main"
};
prototype.__tobelocked = prototype.__tobelocked.concat(["Instantiate", "isa"]);
prototype.__tobefrozen = prototype.__tobefrozen.concat(["isof"]);
var $1g = function() {
  var $ = this;
  return {
    wordcount: 100,
    truncate: undefined,
    window: 3,
    frequency: false,
    weighted: 'AnalyzeUnweighted',
    minimum: undefined,
    markovWord: undefined
  };
}
for (var i in $1g()) {
  prototype[i] = undefined;
};
var $2g = prototype.Constructor || function() {};
prototype.Constructor = function() {
  $2g.call(this);
  var t = $1g();
  for (var i in t)
    if (t[i] !== undefined) this[i] = t[i];
};
var $3g = prototype['Instantiate'] || function() {};
prototype['Instantiate'] = function(p) {
  var $_d, $_c, $_b, $10, $9, $8, $7, $6, $5, _results, $4, _source, $3, $2, $1, $0, _argv, $ = this['Instantiate'] ? this : $bindfail('Instantiate'); {
    $.markovWord = $AI.create_op('MarkovWord', [undefined]);
    _argv = $AI.clone_op(_ARGV);
    if ($0 = (('node' === $AI.slice_op(_argv[0], undefined, (0 - (4)))))) {
      _argv.shift();;
    }
    if ($1 = (('sai-run' === $AI.slice_op(_argv[0], undefined, (0 - (7)))))) {
      _argv.shift();;
    }
    _argv.shift();
    if ($2 = ((_argv.length < 1))) {
      $AI.debug_op('random-words -- Generate new, unique words from a seed set of words' + '\n' + '' + '\n' + 'USAGE' + '\n' + '  random-words [seed.txt]   - print a markov chain generated list of random words' + '\n' + '' + '\n' + 'OPTIONS' + '\n' + '  -c, --count [number]      - maximum number of unique words to generate (default ' + $.wordcount + ')' + '\n' + '  -f, --frequency           - show frequency counts for each unique word' + '\n' + '  -i, --influence           - each input line may have a number, the influence of words on that line' + '\n' + '  -l, --limit [number]      - limit output to how many most frequent words (default all)' + '\n' + '  -m, --minimum [number]    - minimum length of word to generate (default, and lowest, is window size)' + '\n' + '  -w, --window [number]     - size of the markov analysis window in characters, (2+, default ' + $.window + ')' + '\n' + '' + '\n' + 'EXAMPLE' + '\n' + '  random-words -c 50 countries.txt' + '\n' + '' + '\n' + 'NOTE' + '\n' + '  Results are printed to STDOUT, separated by spaces. STDIN is not accepted, you must' + '\n' + '  supply a file. Duplicates, and words that appear in the orginal text are not included' + '\n' + '  in the output. If the probability of generating unique words is low, it may take a while' + '\n' + '  to generate a large set of unique words.' + '\n' + '');
      _Exit();
    }
    var $10;
    while ($10 = (_argv.shift())) {
      switch ($9 = ($10)) {
        case '-c':
        case '--count':
          $.wordcount = ($AI.number_op(_argv.shift()));
          break;
        case '-w':
        case '--window':
          var $3 = ($AI.number_op(_argv.shift()));
          if (!(($3 < 2))) {
            $.window = $3;
          } else {
            $AI.debug_op('Error: --window cannot be less than 2.');
            _Exit();
          };
          break;
        case '-i':
        case '--influence':
          $.weighted = 'AnalyzeWeighted';
          break;
        case '-f':
        case '--frequency':
          $.frequency = true;
          break;
        case '-l':
        case '--limit':
          $.truncate = ($AI.number_op(_argv.shift()));
          break;
        case '-m':
        case '--minimum':
          $.minimum = ($AI.number_op(_argv.shift()));
          break;
        default:
          try {
            _source = _ReadFile($10, 'utf8');
          } catch ($4) {
            $AI.debug_op('Error ' + $4.message + ' reading file.');
            _Exit();
          }
          $.markovWord.Configure($);
          $.markovWord[$.weighted](_source);
          if (!($.markovWord.success)) {
            $AI.debug_op('No input words longer than ' + $.window + ' letters, cannot produce a markov chain. ');
          } else {
            _results = $AI.slice_op($.markovWord.Generate($.wordcount), 0, $.truncate);
            if (!(_results.length)) {
              $AI.debug_op('No unique words generated. Try increasing count or decreasing window size.');
            } else {
              if ($5 = ((_results.length < (((undefined !== ($_b = $.truncate)) ? $_b : $.wordcount) / 2)))) {
                $AI.debug_op('Few unique words generated due to high window size or low input variety.');
              }
              $8 = _results;
              _results = $AI.map_op($8, function($6, $7) {
                return ($AI.slice_op($6, undefined, 1).toUpperCase() + $AI.slice_op($6, 1, undefined));
              });
              $AI.debug_op(_results.join(' '));
            }
          }
          break;
      }
    }
  }
};
$AI.finalizePrototype(prototype);
if (prototype.isof[prototype.isa].type === "singleton") { prototype.Constructor(); prototype.Instantiate(); }
var result=prototype.isof[prototype.isa].type === "main" ? prototype.constructor() : prototype;
exports=result; try { module.exports=result; } catch(e) {};
return result;
